File: vid.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>

int main(int argc, char* argv[]){

  printf("VID hello world\n");

  int eventPipe[2];
  int paintPipe[2];
  pid_t pid;
  char* args[] = {"cor", NULL};
  
  // in both of these pipes, we are pipe 0
  pipe(eventPipe);
  pipe(paintPipe);

  pid = fork();

  if(pid == -1){
    fprintf(stderr, "VID fork failed\n");
    fprintf(stderr, "VID' exec failed! %s\n", strerror(errno));
    exit(-1);
  }
  else if(pid == 0){ // child
    printf("VID' child rearranging file descriptors\n");
    close(eventPipe[0]);
    close(paintPipe[0]);
    close(0);
    dup(eventPipe[1]); // we will read event commands from stdin
    close(1);
    dup(paintPipe[1]); // we will write paint commands to stdout
    execve("cor", args, NULL);
    fprintf(stderr, "VID' exec failed! %s\n", strerror(errno));
    exit(-1);
  }
  else{ // parent
    close(eventPipe[1]);
    close(paintPipe[1]);
    int status;
    wait(&status);
    printf("VID child terminated with code %d\n", status);
  }

  return 0;
}

File: Vid.hs
module Main where

import System.Process
import Control.Monad.Loops (whileM_)
import Data.Functor ((<$>))
import qualified Data.Vector.Storable as V

import qualified Graphics.UI.GLFW as GLFW
import Graphics.GL.Low

-- Spawn the COR process which will react to events written to its
-- stdin. It will also write paint commands which VID is expected to
-- execute.
spawnCore :: IO (Handle, Handle)
spawnCore = do
  (Just paintPipe, Just eventPipe, _, cor) <- createProcess
    (proc "./cor" []) { std_in = CreatePipe, std_out = CreatePipe }
  return (paintPipe, eventPipe)

data PaintCommand =
  PaintBox |
  PaintLine |
  PaintImage |
  PaintText |
  PaintFill
    deriving Show

main = do
  putStrLn "VID hello world"
  (paintIn, eventOut) <- spawnCore
  paintQueue <- newTQueueIO
  forkIO (paintWorker paintQueue)
  glfwLoop eventOut paintQueue

errorCb :: GLFW.Error -> String -> IO ()
errorCb e msg = do
  hPutStrLn stderr (unwords ["VID", show e, msg])
  exitFailure

-- Setup a GLFW window and react to input by writing to the eventPipe.
-- Independently, when you get a paint command, wait up the event listener
-- so that it can render graphics.
glfwMain :: Handle -> TQueue PaintCommand -> IO a
glfwMain eventOut paintQueue = do
  GLFW.setErrorCallback (Just errorCb)
  ok <- GLFW.init
  when (ok == False) $ do
    GLFW.pollEvents
    exitFailure
  GLFW.windowHint (GLFW.WindowHint'ContextVersionMajor 3)
  GLFW.windowHint (GLFW.WindowHint'ContextVersionMinor 2)
  GLFW.windowHint (GLFW.WindowHint'OpenGLForwardCompat True)
  GLFW.windowHint (GLFW.WindowHint'OpenGLProfile GLFW.OpenGLProfile'Core)
  mwin <- GLFW.createWindow 640 480 "Epichord" Nothing Nothing
  case mwin of
    Nothing  -> do
      GLFW.pollEvents
      exitFailure
    Just win -> do
      GLFW.makeContextCurrent (Just win)
      GLFW.swapInterval 1
      setup
      whileM_ (not <$> GLFW.windowShouldClose win) $ do
        glfwLoop eventOut paintQueue

glfwLoop :: Handle -> TQueue PaintCommand -> IO a
glfwLoop eventOut paintQueue = do
  GLFW.pollEvents
  whileM_ (not <$> atomically (isEmptyTQueue paintQueue)) $ do
    command <- readTQueue paintQueue

  GLFW.swapBuffers win

setup = do
  -- establish a VAO
  vao <- newVAO
  bindVAO vao
  -- load shader program
  vsource <- readFile "hello.vert"
  fsource <- readFile "hello.frag"
  prog <- newProgram vsource fsource
  useProgram prog
  -- load vertex data: three 2D vertex positions
  let blob = V.fromList
        [ -0.5, -0.5
        ,    0,  0.5
        ,  0.5, -0.5 ] :: V.Vector Float
  vbo <- newVBO blob StaticDraw
  bindVBO vbo
  -- connect program to vertex data via the VAO
  setVertexLayout [Attrib "position" 2 GLFloat]
  return (vao, prog)

draw vao prog = do
  clearColorBuffer (0,0,0)
  bindVAO vao
  useProgram prog
  drawTriangles 3
